src/sql/src/plan/query.rs

```diff
@@ -58,6 +61,147 @@ use crate::plan::statement::StatementContext;
 use crate::plan::transform_ast;
 use crate::plan::typeconv::{self, CastContext};
 
+// Aug is the type variable assigned to an AST that has already been
+// name-resolved. An AST in this state has global IDs populated next to table
+// names, and local IDs assigned to CTE definitions and references.
+#[derive(Debug, PartialEq, Eq, Hash, Copy, Clone, Default)]
+pub struct Aug;
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub struct ResolvedTableName {
+    id: Id,
+    raw_name: ObjectName,
+}
+
+impl AstDisplay for ResolvedTableName {
+    fn fmt(&self, f: &mut AstFormatter) {
+        f.write_str(format!("[{}: ", self.id));
+        separated(&self.raw_name.0, ".").fmt(f);
+        f.write_str("]");
+    }
+}
+
+impl std::fmt::Display for ResolvedTableName {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.write_str(self.to_ast_string().as_str())
+    }
+}
+
+impl AstInfo for Aug {
+    type Table = ResolvedTableName;
+    type Id = Id;
+}
+
+#[derive(Debug)]
+struct NameResolver<'a, 'b> {
+    qcx: &'a mut QueryContext<'b>,
```

Comment:
Ooh, now that I'm looking at this more closely, it's sorta weirding me out to have the `QueryContext` in here. Like, I feel like it'd be nice to keep the `QueryContext` as a thing that gets passed around only post name resolution, so it doesn't have a `ctes` field that's only used during name resolution and a `built_ctes` field that's only used _post_ name resolution.

Instead I think you could just stick an immutable ref to the catalog in here plus the `ctes` field that currently lives on `QueryContext? And then you can also move the `QueryContext::resolve_table` method over to `NameResolve::resolve_table`. Does that seem workable?